pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HdfsSegmentFetcher.java

```diff
@@ -0,0 +1,131 @@
+import static com.linkedin.pinot.common.utils.CommonConstants.SegmentFetcher.HdfsSegmentFetcher.HADOOP_CONF_PATH;
+
+public class HdfsSegmentFetcher implements SegmentFetcher {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(HdfsSegmentFetcher.class);
+  private FileSystem hadoopFS = null;
+  private int retryCount = 3;
+  private int retryWaitMs = 100;
+
+  @Override
+  public void init(Map<String, String> configs) {
+    String hadoopConfPath = configs.get(HADOOP_CONF_PATH);
+    try {
+      if (configs.containsKey(RETRY)) {
+        retryCount = Integer.valueOf(configs.get(RETRY));
+      }
+      if (configs.containsKey(RETRY_WAITIME_MS)) {
+        retryWaitMs = Integer.valueOf(configs.get(RETRY_WAITIME_MS));
+      }
+      Configuration hadoopConf = getConf(hadoopConfPath);
+      authenticate(hadoopConf, configs);
+      hadoopFS = FileSystem.get(hadoopConf);
+      LOGGER.info("successfully initialized hdfs segment fetcher");
+    } catch (Exception e) {
+      LOGGER.error("failed to initialized the hdfs segment fetcher", e);
+    }
+  }
+
+  private Configuration getConf(String hadoopConfPath) {
+    Configuration hadoopConf = new Configuration();
+    if (Strings.isNullOrEmpty(hadoopConfPath)) {
+      LOGGER.warn("no hadoop conf path is provided, will rely on default config");
+    } else {
+      hadoopConf.addResource(new Path(hadoopConfPath, "core-site.xml"));
+      hadoopConf.addResource(new Path(hadoopConfPath, "hdfs-site.xml"));
+    }
+    return hadoopConf;
+  }
+
+  private void authenticate(Configuration hadoopConf, Map<String, String> configs) {
+    String principal = configs.get(PRINCIPLE);
+    String keytab = configs.get(KEYTAB);
+    if (!Strings.isNullOrEmpty(principal) && !Strings.isNullOrEmpty(keytab)) {
+      UserGroupInformation.setConfiguration(hadoopConf);
+      if (UserGroupInformation.isSecurityEnabled()) {
+        try {
+          if (!UserGroupInformation.getCurrentUser().hasKerberosCredentials()
+              || !UserGroupInformation.getCurrentUser().getUserName().equals(principal)) {
+            LOGGER.info("Trying to authenticate user [%s] with keytab [%s]..", principal, keytab);
+            UserGroupInformation.loginUserFromKeytab(principal, keytab);
+          }
+        }
+        catch (IOException e) {
+          throw new RuntimeException(String.format("Failed to authenticate user principal [%s] with keytab [%s]",
+                  principal, keytab), e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void fetchSegmentToLocal(String uri, File tempFile) throws Exception {
+    LOGGER.debug("starting to fetch segment from hdfs");
+    String tempFilePath = tempFile.getAbsolutePath();
+    try {
+      Path remoteFile = new Path(uri);
+      Path localFile = new Path(tempFile.toURI());
+
+      int attemptCount = 0;
+      while (attemptCount++ < this.retryCount) {
```

Comment:
Would be better to use `RetryPolicy` to implement the retry logic.

Reply:
Done and updated HttpSegmentFetcher with similar logics as well